### PR TITLE
Feature/swm 494 self service manager

### PR DESCRIPTION
**Pull request by simundev**
This feature implements the functionality required for the Self-Service Token Issuance Platform

### CODE REVIEW

File: contracts/factories/SRC20Registry.sol [hunk 1]
```diff
@@ -7,15 +7,22 @@ import "../interfaces/ISRC20Registry.sol";
...
 contract SRC20Registry is ISRC20Registry, Manager {
     using Roles for Roles.Role;
 
-    Roles.Role private _factories;
+    event FactoryAdded(address account);
+    event FactoryRemoved(address account);
+    event SRC20Registered(address token, address tokenOwner);
+    event SRC20Removed(address token);
+    event MinterAdded(address minter);
+    event MinterRemoved(address minter); // @TODO move to interface
```
Comment (tomislavr) [inline contracts/factories/SRC20Registry.sol]: Clear todo by moving event declarations to interface, and some of these are already declared in `ISRC20Registry`.

File: contracts/interfaces/ISRC20.sol [hunk 2]
```diff
@@ -4,7 +4,7 @@ pragma solidity ^0.5.0;
  * @title SRC20 public interface
  */
 interface ISRC20 {
-    function getKYA() external view returns (bytes32, string memory, address);
+    //function getKYA() external view returns (bytes32, string memory, address);
```
Comment (tomislavr) [inline contracts/interfaces/ISRC20.sol]: Remove line.

File: contracts/token/AssetRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+     * @param kyaUrl URL of token's KYA document (ipfs, http, etc.).
+     *               or address(0) if no rules should be checked on chain.
+     * @return True on success.
+     */
+    function addAsset(address src20, bytes32 kyaHash, string calldata kyaUrl, uint256 netAssetValueUSD)
+        external
+        onlyFactory
+        returns (bool)
+    {
+        require(assetList[src20].netAssetValueUSD == 0, 'Asset already added, try update functions');
```
Comment (tomislavr) [inline contracts/token/AssetRegistry.sol]: This `require` is not needed, as only `onlyFactory` can call this and it get called only once - on token creation.

File: contracts/token/SRC20.sol [hunk 4]
```diff
@@ -5,45 +5,43 @@ import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
...
 /**
  * @title SRC20 contract
  * @dev Base SRC20 contract.
  */
-contract SRC20 is ISRC20, ISRC20Owned, ISRC20Managed, SRC20Detailed, Ownable {
+contract SRC20 is ISRC20, ISRC20Managed, SRC20Detailed, Ownable {
     using SafeMath for uint256;
     using ECDSA for bytes32;
 
+    event RestrictionsAndRulesUpdated(address restrictions, address rules);
```
Comment (tomislavr) [inline contracts/token/SRC20.sol]: Maybe move this event declaration to some interface.

File: contracts/token/SRC20.sol [hunk 5]
```diff
@@ -354,6 +338,8 @@ contract SRC20 is ISRC20, ISRC20Owned, ISRC20Managed, SRC20Detailed, Ownable {
     }
 
     function transfer(address to, uint256 value) external returns (bool) {
+        require(!_features.checkTransfer(msg.sender, to));
```
Comment (tomislavr) [inline contracts/token/SRC20.sol]: `checkTransfer(from, to)` has this condition:
```
return _isAccountFrozen(from) && _isAccountFrozen(to) && paused();
```
If `from` is not frozen, all other combinations will result in allowed transfer. Same applies everywhere `checkTransfer` is used.

File: contracts/factories/SRC20Registry.sol (hunk 1 above)
Comment (simundev) [reply contracts/factories/SRC20Registry.sol]: Done

File: contracts/interfaces/ISRC20.sol (hunk 2 above)
Comment (simundev) [reply contracts/interfaces/ISRC20.sol]: Done

File: contracts/token/SRC20.sol (hunk 5 above)
Comment (simundev) [reply contracts/token/SRC20.sol]: Well spotted, thanks. Fix applied.

File: contracts/token/SRC20.sol (hunk 4 above)
Comment (simundev) [reply contracts/token/SRC20.sol]: Moved to ISRC20.sol

File: contracts/token/AssetRegistry.sol (hunk 3 above)
Comment (simundev) [reply contracts/token/AssetRegistry.sol]: True, but Factory could be changed at some point making this function vulnerable. I think it makes sense that this function protects itself from wrong input, so I'll leave it in for now.

**Pull request merged**